commit 1dcd0cf42a6a7cb62cc9a7c024196733abcfda95
Author: Nicholas Wilson <nicholas@nicholaswilson.me.uk>
Date:   Sat Aug 8 19:17:36 2026 +0100

    Fix leak & stale PCRE2_MD_COPIED_SUBJECT if pcre2_jit_match used with existing match context (#937)
    
    The problem is not that pcre2_jit_match() needs to add support for PCRE2_COPY_MATCHED_SUBJECT. Instead, if the passed-in context somehow contains a previously-copied subject (by non-JIT matcher using a global or cached subject) then it will be leaked, and worse, incorrectly free'd later.

diff --git a/doc/html/pcre2jit.html b/doc/html/pcre2jit.html
index cc26cc06..4e6d31e5 100644
--- a/doc/html/pcre2jit.html
+++ b/doc/html/pcre2jit.html
@@ -460,10 +460,11 @@ processed by <b>pcre2_jit_compile()</b>).
 The fast path function is called <b>pcre2_jit_match()</b>, and it takes exactly
 the same arguments as <b>pcre2_match()</b>. However, the subject string must be
 specified with a length; PCRE2_ZERO_TERMINATED is not supported. Unsupported
-option bits (for example, PCRE2_ANCHORED and PCRE2_ENDANCHORED) are ignored, as
-is the PCRE2_NO_JIT option. The return values are also the same as for
-<b>pcre2_match()</b>, plus PCRE2_ERROR_JIT_BADOPTION if a matching mode (partial
-or complete) is requested that was not compiled.
+option bits (for example, PCRE2_ANCHORED, PCRE2_ENDANCHORED, and
+PCRE2_COPY_MATCHED_SUBJECT) are ignored, as is the PCRE2_NO_JIT option. The
+return values are also the same as for <b>pcre2_match()</b>, plus
+PCRE2_ERROR_JIT_BADOPTION if a matching mode (partial or complete) is requested
+that was not compiled.
 </p>
 <p>
 When you call <b>pcre2_match()</b>, as well as testing for invalid options, a
diff --git a/doc/pcre2.txt b/doc/pcre2.txt
index 693908ee..cc316178 100644
--- a/doc/pcre2.txt
+++ b/doc/pcre2.txt
@@ -6194,11 +6194,11 @@ JIT FAST PATH API
        The  fast  path  function is called pcre2_jit_match(), and it takes ex-
        actly the same arguments as pcre2_match(). However, the subject  string
        must  be  specified  with  a  length; PCRE2_ZERO_TERMINATED is not sup-
-       ported.  Unsupported  option  bits  (for  example,  PCRE2_ANCHORED  and
-       PCRE2_ENDANCHORED)  are ignored, as is the PCRE2_NO_JIT option. The re-
-       turn values are also the same  as  for  pcre2_match(),  plus  PCRE2_ER-
-       ROR_JIT_BADOPTION if a matching mode (partial or complete) is requested
-       that was not compiled.
+       ported. Unsupported option bits (for example, PCRE2_ANCHORED, PCRE2_EN-
+       DANCHORED, and  PCRE2_COPY_MATCHED_SUBJECT)  are  ignored,  as  is  the
+       PCRE2_NO_JIT  option.  The  return  values  are  also  the  same as for
+       pcre2_match(), plus PCRE2_ERROR_JIT_BADOPTION if a matching mode  (par-
+       tial or complete) is requested that was not compiled.
 
        When  you call pcre2_match(), as well as testing for invalid options, a
        number of other sanity checks are performed on the arguments. For exam-
diff --git a/doc/pcre2jit.3 b/doc/pcre2jit.3
index 95451b56..729d898f 100644
--- a/doc/pcre2jit.3
+++ b/doc/pcre2jit.3
@@ -444,10 +444,11 @@ processed by \fBpcre2_jit_compile()\fP).
 The fast path function is called \fBpcre2_jit_match()\fP, and it takes exactly
 the same arguments as \fBpcre2_match()\fP. However, the subject string must be
 specified with a length; PCRE2_ZERO_TERMINATED is not supported. Unsupported
-option bits (for example, PCRE2_ANCHORED and PCRE2_ENDANCHORED) are ignored, as
-is the PCRE2_NO_JIT option. The return values are also the same as for
-\fBpcre2_match()\fP, plus PCRE2_ERROR_JIT_BADOPTION if a matching mode (partial
-or complete) is requested that was not compiled.
+option bits (for example, PCRE2_ANCHORED, PCRE2_ENDANCHORED, and
+PCRE2_COPY_MATCHED_SUBJECT) are ignored, as is the PCRE2_NO_JIT option. The
+return values are also the same as for \fBpcre2_match()\fP, plus
+PCRE2_ERROR_JIT_BADOPTION if a matching mode (partial or complete) is requested
+that was not compiled.
 .P
 When you call \fBpcre2_match()\fP, as well as testing for invalid options, a
 number of other sanity checks are performed on the arguments. For example, if
diff --git a/src/pcre2_jit_match_inc.h b/src/pcre2_jit_match_inc.h
index 32d4c8a5..4163cf61 100644
--- a/src/pcre2_jit_match_inc.h
+++ b/src/pcre2_jit_match_inc.h
@@ -125,6 +125,16 @@ else if ((options & PCRE2_PARTIAL_SOFT) != 0)
 if (functions == NULL || functions->executable_funcs[index] == NULL)
   return match_data->rc = PCRE2_ERROR_JIT_BADOPTION;
 
+/* If the match data block was previously used with PCRE2_COPY_MATCHED_SUBJECT,
+free the memory that was obtained. */
+
+if ((match_data->flags & PCRE2_MD_COPIED_SUBJECT) != 0)
+  {
+  match_data->memctl.free((void *)match_data->subject,
+    match_data->memctl.memory_data);
+  match_data->flags &= ~PCRE2_MD_COPIED_SUBJECT;
+  }
+
 /* Sanity checks should be handled by pcre2_match. */
 arguments.str = subject + start_offset;
 arguments.begin = subject;
diff --git a/src/pcre2test_inc.h b/src/pcre2test_inc.h
index 5d282435..a74e3368 100644
--- a/src/pcre2test_inc.h
+++ b/src/pcre2test_inc.h
@@ -5199,20 +5199,28 @@ for (gmatched = 0;; gmatched++)
     /* If PCRE2_COPY_MATCHED_SUBJECT was set, check that things are as they
     should be, but not for fast JIT, where it isn't supported. */
 
-    if ((dat_datctl.options & PCRE2_COPY_MATCHED_SUBJECT) != 0 &&
-        (pat_patctl.control & CTL_JITFAST) == 0)
+    if ((dat_datctl.options & PCRE2_COPY_MATCHED_SUBJECT) != 0)
       {
-      if ((match_data->flags & PCRE2_MD_COPIED_SUBJECT) == 0)
-        cfprintf(clr_test_error, outfile,
-          "** PCRE2 error: flag not set after copy_matched_subject\n");
+      if ((pat_patctl.control & CTL_JITFAST) != 0)
+        {
+        if ((match_data->flags & PCRE2_MD_COPIED_SUBJECT) != 0)
+          cfprintf(clr_test_error, outfile,
+            "** PCRE2 error: flag set after unsupported copy_matched_subject\n");
+        }
+      else
+        {
+        if ((match_data->flags & PCRE2_MD_COPIED_SUBJECT) == 0)
+          cfprintf(clr_test_error, outfile,
+            "** PCRE2 error: flag not set after copy_matched_subject\n");
 
-      if (match_data->subject == pp)
-        cfprintf(clr_test_error, outfile,
-          "** PCRE2 error: copy_matched_subject has not copied\n");
+        if (match_data->subject == pp)
+          cfprintf(clr_test_error, outfile,
+            "** PCRE2 error: copy_matched_subject has not copied\n");
 
-      if (memcmp(match_data->subject, pp, ulen) != 0)
-        cfprintf(clr_test_error, outfile,
-          "** PCRE2 error: copy_matched_subject mismatch\n");
+        if (memcmp(match_data->subject, pp, ulen) != 0)
+          cfprintf(clr_test_error, outfile,
+            "** PCRE2 error: copy_matched_subject mismatch\n");
+        }
       }
 
     /* If this is not the first time round a global loop, check that the
@@ -5689,6 +5697,9 @@ pcre2_match_context *test_dat_context = NULL, *test_dat_context_copy = NULL;
 pcre2_convert_context *test_con_context = NULL, *test_con_context_copy = NULL;
 pcre2_match_data *test_match_data = NULL;
 pcre2_code *test_compiled_code = NULL;
+#ifdef SUPPORT_JIT
+BOOL test_compiled_with_jit = FALSE;
+#endif
 PCRE2_UCHAR pattern[] = { CHAR_A, CHAR_B, CHAR_C, 0 };
 PCRE2_UCHAR callout_int_pattern[] = {
   CHAR_LEFT_PARENTHESIS, CHAR_QUESTION_MARK, CHAR_C, CHAR_RIGHT_PARENTHESIS, 0 };
@@ -5993,11 +6004,38 @@ ASSERT(rc == 0 && sizeval == 0, "pcre2_pattern_info(JIT)");
 
 if (pcre2_jit_compile(test_compiled_code, PCRE2_JIT_COMPLETE) == 0)
   {
+  test_compiled_with_jit = TRUE;
+
   rc = pcre2_pattern_info(test_compiled_code, PCRE2_INFO_JITSIZE, &sizeval);
   ASSERT(rc == 0 && sizeval > 0, "pcre2_pattern_info(JIT after compile)");
   }
 #endif
 
+/* ----------------------- Matching functions ------------------------------ */
+
+#ifdef SUPPORT_JIT
+
+/* Check that fast JIT releases a copied subject when reusing match data. */
+if (test_compiled_with_jit)
+  {
+  test_match_data = pcre2_match_data_create_from_pattern(test_compiled_code,
+    test_gen_context);
+  ASSERT(test_match_data != NULL, "pcre2_match_data_create_from_pattern(JIT)");
+
+  rc = pcre2_match(test_compiled_code, pattern, 3, 0,
+    PCRE2_COPY_MATCHED_SUBJECT, test_match_data, NULL);
+  ASSERT(rc == 1, "pcre2_match(COPY_MATCHED_SUBJECT)");
+
+  rc = pcre2_jit_match(test_compiled_code, subject_abcz, 4, 0, 0,
+    test_match_data, NULL);
+  ASSERT(rc == 1, "pcre2_jit_match(reused match data)");
+
+  pcre2_match_data_free(test_match_data);
+  test_match_data = NULL;
+  }
+
+#endif
+
 /* ----------------------- POSIX functions --------------------------------- */
 
 #if PCRE2_CODE_UNIT_WIDTH == 8
diff --git a/testdata/testinput17 b/testdata/testinput17
index 08fd72e0..9d728965 100644
--- a/testdata/testinput17
+++ b/testdata/testinput17
@@ -298,6 +298,7 @@
     
 /abc/jitfast
     abc
+    abc\=copy_matched_subject
     abc\=no_jit 
     
 # ---- 
diff --git a/testdata/testoutput17 b/testdata/testoutput17
index 6d550084..773ec18a 100644
--- a/testdata/testoutput17
+++ b/testdata/testoutput17
@@ -542,6 +542,8 @@ Failed: error -47: match limit exceeded
     
 /abc/jitfast
     abc
+ 0: abc (JIT)
+    abc\=copy_matched_subject
  0: abc (JIT)
     abc\=no_jit 
  0: abc (JIT)
commit 1cad831209af67d24298c301a74c615d3a6d57eb
Author: Carlo Marcelo Arenas Belón <carenas@gmail.com>
Date:   Sat Aug 22 04:02:30 2026 -0700

    pcre2test: honor no_jit when used with jitfast (#946)
    
    `pcre2_jit_match()` ignores the PCRE2_NO_JIT option, so teach pcre2test
    to avoid calling it if the subject modifier that sets that option is
    used together with the option to call it directly.
    
    While at it, make jitverify more reliable and add a unittest to validate
    that PCRE2_NO_JIT is ignored in the fast JIt path.

diff --git a/doc/html/pcre2test.html b/doc/html/pcre2test.html
index 5160acc9..d2534780 100644
--- a/doc/html/pcre2test.html
+++ b/doc/html/pcre2test.html
@@ -863,7 +863,7 @@ when it can never be used.
 The <b>framesize</b> modifier shows the size, in bytes, of each storage frame
 used by <b>pcre2_match()</b> for handling backtracking. The size depends on the
 number of capturing parentheses in the pattern. A vector of these frames is
-used at matching time; its overall size is shown when the <b>heaframes_size</b>
+used at matching time; its overall size is shown when the <b>heapframes_size</b>
 subject modifier is set.
 </p>
 <p>
diff --git a/doc/pcre2test.1 b/doc/pcre2test.1
index 897552a2..c5bc9317 100644
--- a/doc/pcre2test.1
+++ b/doc/pcre2test.1
@@ -811,7 +811,7 @@ when it can never be used.
 The \fBframesize\fP modifier shows the size, in bytes, of each storage frame
 used by \fBpcre2_match()\fP for handling backtracking. The size depends on the
 number of capturing parentheses in the pattern. A vector of these frames is
-used at matching time; its overall size is shown when the \fBheaframes_size\fP
+used at matching time; its overall size is shown when the \fBheapframes_size\fP
 subject modifier is set.
 .P
 The \fBcallout_info\fP modifier requests information about all the callouts in
diff --git a/doc/pcre2test.txt b/doc/pcre2test.txt
index 819fe76e..fad3b8f3 100644
--- a/doc/pcre2test.txt
+++ b/doc/pcre2test.txt
@@ -787,7 +787,7 @@ PATTERN MODIFIERS
        used by pcre2_match() for handling backtracking. The  size  depends  on
        the  number  of capturing parentheses in the pattern. A vector of these
        frames is used at matching time; its overall size  is  shown  when  the
-       heaframes_size subject modifier is set.
+       heapframes_size subject modifier is set.
 
        The  callout_info  modifier requests information about all the callouts
        in the pattern. A list of them is output at the end of any other infor-
diff --git a/src/pcre2test_inc.h b/src/pcre2test_inc.h
index 72767916..69c64f73 100644
--- a/src/pcre2test_inc.h
+++ b/src/pcre2test_inc.h
@@ -3165,7 +3165,8 @@ for (;;)
       dat_context, dfa_workspace, DFA_WS_DIMENSION);
     }
 
-  else if ((pat_patctl.control & CTL_JITFAST) != 0)
+  else if ((pat_patctl.control & CTL_JITFAST) != 0 &&
+           (dat_datctl.options & PCRE2_NO_JIT) == 0)
     capcount = pcre2_jit_match(compiled_code, pp, ulen, dat_datctl.offset,
       dat_datctl.options, match_data, dat_context);
 
@@ -4747,7 +4748,8 @@ if (dat_datctl.replacement[0] != MOD_STR_UNSET)
         dat_datctl.offset, dat_datctl.options, match_data,
         use_dat_context, dfa_workspace, DFA_WS_DIMENSION);
       }
-    else if ((pat_patctl.control & CTL_JITFAST) != 0)
+    else if ((pat_patctl.control & CTL_JITFAST) != 0 &&
+             (dat_datctl.options & PCRE2_NO_JIT) == 0)
       {
       (void)pcre2_jit_match(compiled_code, pp, arg_ulen, dat_datctl.offset,
         dat_datctl.options, match_data, use_dat_context);
@@ -4990,7 +4992,7 @@ for (gmatched = 0;; gmatched++)
   /* When matching is via pcre2_match(), we will detect the use of JIT via the
   stack callback function. */
 
-  jit_was_used = (pat_patctl.control & CTL_JITFAST) != 0;
+  jit_was_used = FALSE;
 
   /* Do timing if required. */
 
@@ -5021,7 +5023,8 @@ for (gmatched = 0;; gmatched++)
         }
       }
 
-    else if ((pat_patctl.control & CTL_JITFAST) != 0)
+    else if ((pat_patctl.control & CTL_JITFAST) != 0 &&
+             (dat_datctl.options & PCRE2_NO_JIT) == 0)
       {
       start_time = clock();
       for (i = 0; i < timeitm; i++)
@@ -5105,9 +5108,11 @@ for (gmatched = 0;; gmatched++)
       }
     else
       {
-      if ((pat_patctl.control & CTL_JITFAST) != 0)
-        capcount = pcre2_jit_match(compiled_code, pp, arg_ulen, dat_datctl.offset,
-          dat_datctl.options | g_notempty, match_data, use_dat_context);
+      if ((pat_patctl.control & CTL_JITFAST) != 0 &&
+          (dat_datctl.options & PCRE2_NO_JIT) == 0)
+        capcount = pcre2_jit_match(compiled_code, pp, arg_ulen,
+          dat_datctl.offset, dat_datctl.options | g_notempty, match_data,
+          use_dat_context);
       else
         capcount = pcre2_match(compiled_code, pp, arg_ulen, dat_datctl.offset,
           dat_datctl.options | g_notempty, match_data, use_dat_context);
@@ -5142,12 +5147,15 @@ for (gmatched = 0;; gmatched++)
           }
         else
           {
-          if ((pat_patctl.control & CTL_JITFAST) != 0)
-            capcount = pcre2_jit_match(compiled_code, pp, arg_ulen, dat_datctl.offset,
-              dat_datctl.options | g_notempty, match_data, use_dat_context);
+          if ((pat_patctl.control & CTL_JITFAST) != 0 &&
+              (dat_datctl.options & PCRE2_NO_JIT) == 0)
+            capcount = pcre2_jit_match(compiled_code, pp, arg_ulen,
+              dat_datctl.offset, dat_datctl.options | g_notempty, match_data,
+              use_dat_context);
           else
-            capcount = pcre2_match(compiled_code, pp, arg_ulen, dat_datctl.offset,
-              dat_datctl.options | g_notempty, match_data, use_dat_context);
+            capcount = pcre2_match(compiled_code, pp, arg_ulen,
+              dat_datctl.offset, dat_datctl.options | g_notempty, match_data,
+              use_dat_context);
           }
 
         mallocs_until_failure = INT_MAX;
@@ -5207,12 +5215,20 @@ for (gmatched = 0;; gmatched++)
       return PR_ABEND;
       }
 
+    /* tracking JIT with jitverify needs a context, but in cases where one
+    wasn't available, can fallback to the match_data status */
+
+    if ((dat_datctl.control & CTL_NULLCONTEXT) != 0 &&
+        (pat_patctl.control & CTL_JITVERIFY) != 0)
+      jit_was_used = match_data->matchedby == PCRE2_MATCHEDBY_JIT;
+
     /* If PCRE2_COPY_MATCHED_SUBJECT was set, check that things are as they
     should be, but not for fast JIT, where it isn't supported. */
 
     if ((dat_datctl.options & PCRE2_COPY_MATCHED_SUBJECT) != 0)
       {
-      if ((pat_patctl.control & CTL_JITFAST) != 0)
+      if ((pat_patctl.control & CTL_JITFAST) != 0 &&
+          (dat_datctl.options & PCRE2_NO_JIT) == 0)
         {
         if ((match_data->flags & PCRE2_MD_COPIED_SUBJECT) != 0)
           cfprintf(clr_test_error, outfile,
@@ -6026,9 +6042,11 @@ if (pcre2_jit_compile(test_compiled_code, PCRE2_JIT_COMPLETE) == 0)
 
 #ifdef SUPPORT_JIT
 
-/* Check that fast JIT releases a copied subject when reusing match data. */
 if (test_compiled_with_jit)
   {
+
+/* Check that fast JIT releases a copied subject when reusing match data. */
+
   test_match_data = pcre2_match_data_create_from_pattern(test_compiled_code,
     test_gen_context);
   ASSERT(test_match_data != NULL, "pcre2_match_data_create_from_pattern(JIT)");
@@ -6043,6 +6061,22 @@ if (test_compiled_with_jit)
 
   pcre2_match_data_free(test_match_data);
   test_match_data = NULL;
+
+/* Check that fast JIT ignores PCRE2_NO_JIT */
+
+  test_match_data = pcre2_match_data_create(100000, test_gen_context);
+  ASSERT(test_match_data != NULL, "pcre2_match_data_create(100000)");
+  ASSERT(pcre2_get_ovector_count(test_match_data) == 65535,
+    "pcre2_get_ovector_count(UINT32_MAX) <= UINT16_MAX)");
+
+  rc = pcre2_jit_match(test_compiled_code, subject_abcz, 4, 0, PCRE2_NO_JIT,
+    test_match_data, NULL);
+  ASSERT(rc == 1 && (test_match_data->matchedby == PCRE2_MATCHEDBY_JIT),
+    "pcre2_jit_match(ignore PCRE2_NO_JIT");
+
+  pcre2_match_data_free(test_match_data);
+  test_match_data = NULL;
+
   }
 
 #endif
diff --git a/testdata/testinput17 b/testdata/testinput17
index e6dbbe42..a02e6be2 100644
--- a/testdata/testinput17
+++ b/testdata/testinput17
@@ -295,12 +295,29 @@
 /abc/
     abc
     abc\=no_jit 
-    
+
+/(*NO_JIT)abc/
+    abc
+
 /abc/jitfast
     abc
-    abc\=copy_matched_subject
     abc\=no_jit 
-    
+
+/(*NO_JIT)abc/jitfast
+\= Expect error
+    abc
+
+# Do not leak a previously copied subject with jitfast (#920)
+
+/abc/jitfast
+    abc\=copy_matched_subject,no_jit
+    abc
+
+/abc/jitfast
+    abc\=copy_matched_subject
+    abc\=null_context
+    abc\=null_context,no_jit
+
 # ---- 
 
 /[aC]/mg,firstline,newline=lf
diff --git a/testdata/testoutput17 b/testdata/testoutput17
index 11f032d8..c678587f 100644
--- a/testdata/testoutput17
+++ b/testdata/testoutput17
@@ -539,15 +539,38 @@ Failed: error -47: match limit exceeded
  0: abc (JIT)
     abc\=no_jit 
  0: abc
-    
+
+/(*NO_JIT)abc/
+    abc
+ 0: abc
+
+/abc/jitfast
+    abc
+ 0: abc (JIT)
+    abc\=no_jit 
+ 0: abc
+
+/(*NO_JIT)abc/jitfast
+\= Expect error
+    abc
+Failed: error -45: bad JIT option
+
+# Do not leak a previously copied subject with jitfast (#920)
+
 /abc/jitfast
+    abc\=copy_matched_subject,no_jit
+ 0: abc
     abc
  0: abc (JIT)
+
+/abc/jitfast
     abc\=copy_matched_subject
  0: abc (JIT)
-    abc\=no_jit 
+    abc\=null_context
  0: abc (JIT)
-    
+    abc\=null_context,no_jit
+ 0: abc
+
 # ---- 
 
 /[aC]/mg,firstline,newline=lf
